### PR TITLE
[IBCDPE-852] Update filenames before storing to Synapse

### PR DESCRIPTION
**Pull request by jaymedina**
### problem

This is a preventative fix for a potential edge-case in model-to-data Challenge submissions.

The edge-case involves the potential processing of multiple submissions from the same submitter: Currently the infrastructure uploads the generically named predictions files to Synapse, before changing their metadata to reflect a name more specific to the submission ID. This can cause a potential collision if more than one predictions file (`predictions.csv/zip`) is being uploaded into Synapse before the previous predictions file has been renamed.

### solution

The solution to this edge case involves renaming the output files, with their unique submission IDs, before upload to Synapse, so that there is no clash between submissions from the same submitter. This solution involves:

- [x] The creation of a new NF script called `rename_file.nf`
- [x] The creation of a new python script called `rename_file.py`
- [x] The model-to-data workflow incorporating the new `RENAME_FILE` process

### testing & preview

* Tower run with 5 submissions from the same submitter all at once: https://tower.sagebionetworks.org/orgs/Sage-Bionetworks/workspaces/pegs-challenge-project/watch/1UtoH4UVwKcq2F

* Resulting output files:

image

* An example execution log for `rename_file.nf`:

image

----
### Notes:
I am not going to update documentation in this PR, as this feature branch is going to be merged into another feature branch, which is where I _will_ be updating the documentation. Please review this PR keeping in mind that it will be merged into another feature branch, and has been tracking its diff off that other feature branch (not main), so any inconsistencies with `main` is due to that.

### CODE REVIEW

**Comment by jaymedina** [comment]
Note: I've sent this back into draft mode after discovering the predictions files for submissions 9743264 and -65 were swapped (see execution logs of [this run](https://tower-dev.sagebionetworks.org/orgs/Sage-Bionetworks/workspaces/example-dev-project/watch/3NcCICcbOwTRzg)). Investigating.

image

File: workflows/MODEL_TO_DATA.nf [hunk 1]
```diff
@@ -46,13 +47,14 @@ workflow MODEL_TO_DATA {
...
-    UPDATE_FOLDERS(submission_ch, params.project_name, RUN_DOCKER.output.map { it[1] }, RUN_DOCKER.output.map { it[2] })
-    // UPDATE_SUBMISSION_STATUS_AFTER_RUN(RUN_DOCKER.output.map { it[0] }, "ACCEPTED")
-    // VALIDATE(RUN_DOCKER.output, UPDATE_SUBMISSION_STATUS_AFTER_RUN.output, params.validation_script)
-    // UPDATE_SUBMISSION_STATUS_AFTER_VALIDATE(submission_ch, VALIDATE.output.map { it[2] })
-    // ANNOTATE_SUBMISSION_AFTER_VALIDATE(VALIDATE.output)
-    // SCORE(VALIDATE.output, UPDATE_SUBMISSION_STATUS_AFTER_VALIDATE.output, ANNOTATE_SUBMISSION_AFTER_VALIDATE.output, params.scoring_script)
-    // UPDATE_SUBMISSION_STATUS_AFTER_SCORE(submission_ch, SCORE.output.map { it[2] })
-    // ANNOTATE_SUBMISSION_AFTER_SCORE(SCORE.output)
-    // SEND_EMAIL(params.view_id, submission_ch, params.email_with_score, ANNOTATE_SUBMISSION_AFTER_SCORE.output)
+    renamed_file = RENAME_FILE(submission_ch, RUN_DOCKER.output.map { it[1] })
```
Comment (jaymedina) [inline workflows/MODEL_TO_DATA.nf]: Something interesting happened in one of my latest runs (5 submissions per submitter) where the predictions files were uploaded on behalf of another submission. See the command line call for the UPDATE_FOLDERS task for submissions 9743264 and 9743264 for context: [this run](https://tower-dev.sagebionetworks.org/orgs/Sage-Bionetworks/workspaces/example-dev-project/watch/3NcCICcbOwTRzg)).

Technically this swap doesn't do any harm (the predictions files all end up in the same place at the end of the day), but it does call for attention the way the output values are fed into subsequent processes. My solution was to explicitly assign a variable to the output file and feeding that new variable (`renamed_file`) to `UPDATE_FOLDERS`, instead of implicitly dumping `RENAME_FILE.output` into `UPDATE_FOLDERS`.

Comment (jaymedina) [reply workflows/MODEL_TO_DATA.nf]: [New run with the fix](https://tower-dev.sagebionetworks.org/orgs/Sage-Bionetworks/workspaces/example-dev-project/watch/5T1584ulSs24so)

Comment (BWMac) [reply workflows/MODEL_TO_DATA.nf]: This is very strange, were you able to determine why that was happening?

**Review by BWMac — COMMENTED** [review]
What is your rationale for creating a new process to do the renaming, rather than doing the renaming as part of `RUN_DOCKER`? Are you anticipating a case where we might want to have the predictions files without the submission id appended? Especially given that the log files already come out of `RUN_DOCKER` with the correct name.

Edit: For example, you could append `"{submission_id}_"` to any file in the `output` folder in `run_docker.py` instead of setting the `docker.log` file name [here](https://github.com/Sage-Bionetworks-Workflows/nf-synapse-challenge/blob/761b4cc79a56354e51f97dd411128be4087ae462/bin/run_docker.py#L186).

File: workflows/MODEL_TO_DATA.nf (hunk 1 above)
Comment (jaymedina) [reply workflows/MODEL_TO_DATA.nf]: Prior to this change, I was feeding `RENAME_FILE.output` to `UPDATE_FOLDERS` which from my understanding is basically the equivalent of telling Nextflow "just run `UPDATE_FOLDERS` with the first output channel of `RENAME_FILE` that you get".

After looking at the submit/start/stop times of the processes for submissions `9743264` and `9743265` in [the workflow where the swap happened](https://tower-dev.sagebionetworks.org/orgs/Sage-Bionetworks/workspaces/example-dev-project/watch/3NcCICcbOwTRzg), I'm seeing that `RENAME_FILE` for -65 finished 1 second faster than for -64:

`9743264`
image

`9743265`
image

This leads me to think that it was an issue of the first output channel for RENAME_FILE being the one that was grabbed for -64, and that happened to be the output path for -65's `predictions` file. But then it leaves me asking how this mixup doesn't happen for things like the `docker.log` files which don't seem to have any mixups (for my test1-5 submissions I have different stdouts that get written to the docker logs to verify this). I'm still trying to grasp the concept of channels.

For what it's worth: in the [new run](https://tower-dev.sagebionetworks.org/orgs/Sage-Bionetworks/workspaces/example-dev-project/watch/5T1584ulSs24so), RENAME_FILE for -65 doesn't start until -64 ends, which prevents this issue from happening. [Here](https://tower-dev.sagebionetworks.org/orgs/Sage-Bionetworks/workspaces/example-dev-project/watch/2RHAGSG6w8JlqD) is another run with the latest branch that ran as expected and shows the same behavior.

**Comment by jaymedina** [comment]
> What is your rationale for creating a new process to do the renaming, rather than doing the renaming as part of RUN_DOCKER?

Renaming a file is a very general piece of functionality that I could see having multiple use-cases, not just to prep a file before upload to Synapse. Later down the line I was thinking of generalizing it even further so that it's not just prefixing a filename, but renaming a file in different ways with different attributes beyond just the submission ID.

**Comment by BWMac** [comment]
Having this split up may not be necessary for now if there is no reason the predictions file produced by RUN_DOCKER should not have the ID appended to the file name. As it stands, you can't use RUN_DOCKER without RENAME_FILE for any of our current use cases if I am understanding this correctly. It's also not ideal that RUN_DOCKER produces one file with the ID appended, and one without it IMHO (Edit: This is the main thing that rubs me the wrong way with this approach). So I recommend folding this functionality in with RUN_DOCKER.

**Comment by jaymedina** [comment]
Wasn't the goal to stick to a fully modular set of tasks and not fold in any hidden functionality to a particular task but rather have multiple smaller-scoped tasks? I liked that philosophy and am in agreement with it. We may not always have the Logs output structure follow the one I proposed (`predictons/[submission-id]_predictions.csv/zip`), so folding that into `RUN_DOCKER` will age it badly if things change. Once I'm back from break I can think of a middle ground, but I'm not a fan of folding anything into `RUN_DOCKER` at the moment, based on the architecture discussion last week. I'll need to think about this more.

**Comment by BWMac** [comment]
So I guess the disagreement is that I don't feel that this approach is compatible with the modular philosophy. To me, modular doesn't mean that processes are as limited in scope as possible, but rather that they have clearly defined goals that they accomplish. With the goals of our current use case in mind,  I see no reason why `RUN_DOCKER` should produce a logs file with the ID and another file with no ID.  This means that the output of `RUN_DOCKER` then needs to be split up and partially run through `RENAME_FILE` and partially passed straight to `UPDATE_FOLDERS`. As it stands, I see `RUN_DOCKER` as an incomplete process, missing the functionality of naming the output files properly. 

Additionally, if the output folder structure changes, `RUN_DOCKER` will need further changes anyway.

Happy to discuss more when you return from vacation.

**Comment by BryanFauble** [comment]
+1 on @BWMac Thoughts. I think it's important to dissect when something should be a utility function, and when it should be a separate process or step in the pipeline.

IMO A rename function should be a utility function that is added to a common spot, that is re-usable, but not necessarily another step in a pipeline.

The process where the file is produced should do some encapsulation and handle the specifics of how that process intends the outside world to view and reference the file it produces.

**Comment by jaymedina** [comment]
Hey all, I'm back! @BryanFauble I like the idea of having a set of utility/helper functions used to handle specific pieces of logic for a script in a given task. This was the thinking I had when I created `helpers.py` in https://github.com/Sage-Bionetworks-Workflows/nf-synapse-challenge/pull/16. The drawback to this approach is that, once again, if we work on a Challenge with a different architecutre for the `Logs/` folder (e.g. we go back to storing predictions files with the generic `predictions.csv/zip` names), `run_docker.py` would have to change, too. However, the change would be minimal, since it's just a matter of removing an import and function call. If @BWMac is also in agreement with turning this into a utility function, I am onboard with it and can do this in the next commit. 

TLDR: I like the idea of converting `RENAME_FILE` into a utility function that then gets called into `run_docker.py`. This introduces the _possibility_ of tech debt, if for some reason or another we no longer require the predictions files to be renamed, but the tech debt is minimal (to me). 

----

>  _To me, modular doesn't mean that processes are as limited in scope as possible, but rather that they have clearly defined goals that they accomplish._

Thanks @BWMac for sharing this, I think I see your train of thought better. For me, the definition of modularity is that a task is *general* enough that it can be implemented in many contexts (i.e. PEGS Challenge, Dynamic Challenge, etc) without the need to modify it. In other words, the ideal task is designed so that it can be plugged into any workflow and work out of the box with the correct inputs and outputs, without having to modify the task any further. I think this is an important conversation we're having, and the architecture is something we can continue iterating on as we continue to "sculpt" this repository and see it used in different challenges.

**Comment by BWMac** [comment]
@jaymedina Sounds good to me!  

> For me, the definition of modularity is that a task is general enough that it can be implemented in many contexts (i.e. PEGS > Challenge, Dynamic Challenge, etc) without the need to modify it.

I agree with this 100% and I don't think our perspectives are in conflict when it comes to modularity. My idea has always been that we should be striving to create processes/tasks that represent a complete piece of a challenge workflow AND can be used in multiple contexts which is tricky without knowing future use cases.

Another thing I was thinking about this morning is that Nextflow processes should be considered to be relatively "high-level" pieces of the overall workflow. We probably don't want to have a process that is as trivial as just renaming a file because each process that we run on Tower requires waiting for wall time, provisioning an EC2 instance, pulling down a docker container, and running the process. Where it makes sense, small pieces of logic should be included in existing processes for efficiency purposes as well. Or, if it doesn't make sense for a given situation to fold the logic into an existing process we can do trivial operations like file renaming with bash script logic instead of introducing a new Python script into the mix.

**Comment by jaymedina** [comment]
Hi all, thanks for your feedback. Linked in the description and [here](https://tower.sagebionetworks.org/orgs/Sage-Bionetworks/workspaces/pegs-challenge-project/watch/1UtoH4UVwKcq2F) is the newest workflow run pointed at this latest version of the branch after `RENAME_FILE` has been converted to a helper function in `helpers.py` and pulled into `docker.py` directly.

Thanks!

**Review by BryanFauble — APPROVED** [review]
LGTM!

File: bin/run_docker.py [hunk 2]
```diff
@@ -180,6 +185,21 @@ def run_docker(submission_id: str, log_file_name: str = "docker.log") -> None:
...
+    if rename_output:
+        # Get the output directory based on the mounted volumes dictionary used to run the container
+        output_dir = next((key for key in volumes.keys() if "output" in volumes[key]["bind"]), None)
+        if not os.path.exists(output_dir):
+            raise ValueError("Output directory found in ``volumes`` does not exist.")
+
+        # Get the predictions file, assuming the file exists. Otherwise, ``predictions_file = None``
+        predictions_file = next((file for file in glob(os.path.join(output_dir, "predictions.*")) if os.path.exists(file)), None)
+        if predictions_file is None:
+            raise ValueError("No predictions file found in output directory.")
```
Comment (BWMac) [inline bin/run_docker.py:198]: Isn't this something that we handle in the validate step? If we throw an error here will the submitter be notified of the issue?

**Review by BWMac — APPROVED** [review]
Looks good! Just one comment/question.

File: bin/run_docker.py (hunk 2 above)
Comment (BWMac) [reply bin/run_docker.py:198]: Although I guess the Nextflow process also would fail if the predictions file were missing...

Comment (jaymedina) [reply bin/run_docker.py:198]: I'll merge this and continue this conversation in the other PR.
